engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/StaticAsOfJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,390 @@
+            if (isLeftSide) {
+                decorateLeftSide(chunkOk, sourceKeyChunks, hashSlots, hashOffset, foundBuilder);
+            } else {
+                decorateWithRightSide(chunkOk, sourceKeyChunks);
+            }
+        }
+    }
+
+    @Override
+    public void probeLeft(RowSequence leftRowSet, ColumnSource<?>[] leftSources) {
+        if (leftRowSet.isEmpty()) {
+            return;
+        }
+        try (final ProbeContext pc = makeProbeContext(leftSources, leftRowSet.size())) {
+            probeTable(pc, leftRowSet, false, leftSources, new StaticProbeHandler(true, null, null, null));
+        }
+    }
+
+    @Override
+    public int probeLeft(RowSequence leftRowSet, ColumnSource<?>[] leftSources, @NotNull final LongArraySource slots,
+            RowSetBuilderRandom foundBuilder) {
+        if (leftRowSet.isEmpty()) {
+            return 0;
+        }
+        try (final ProbeContext pc = makeProbeContext(leftSources, leftRowSet.size())) {
+            final MutableInt slotCount = new MutableInt();
+            probeTable(pc, leftRowSet, false, leftSources,
+                    new StaticProbeHandler(true, slots, slotCount, foundBuilder));
+            return slotCount.intValue();
+        }
+    }
+
+    @Override
+    public void probeRight(RowSequence rightRowSet, ColumnSource<?>[] rightSources) {
+        if (rightRowSet.isEmpty()) {
+            return;
+        }
+        try (final ProbeContext pc = makeProbeContext(rightSources, rightRowSet.size())) {
+            probeTable(pc, rightRowSet, false, rightSources, new StaticProbeHandler(false, null, null, null));
+        }
+    }
+
+    @Override
+    public int getTableSize() {
+        return tableSize;
+    }
+
+    @Override
+    public int getOverflowSize() {
+        return 0;
+    }
+
+    /**
+     * When we get the left RowSet out of our source (after a build or probe); we do it by pulling a sequential builder
+     * and then calling build(). We also null out the value in the column source, thus freeing the builder's memory.
+     *
+     * This also results in clearing out the left hand side of the table between each probe phase for the left
+     * refreshing case.
+     *
+     * @param slot the slot in the table
+     * @return the RowSet for this slot
+     */
+    @Override
+    public RowSet getLeftIndex(long slot) {
+        RowSetBuilderSequential builder = (RowSetBuilderSequential) leftRowSetSource.getAndSetUnsafe(slot, null);
+        if (builder == null) {
+            return null;
+        }
+        return builder.build();
+    }
+
+    @Override
+    public RowSet getRightIndex(long slot) {
+        if (rightBuildersConverted) {
+            return (RowSet) rightRowSetSource.getUnsafe(slot);
+        }
+        RowSetBuilderSequential builder = (RowSetBuilderSequential) rightRowSetSource.getUnsafe(slot);
+        if (builder == null) {
+            return null;
+        }
+        return builder.build();
```

Comment:
I don't think a builder can give you an index more than once; so it might actually be the right move to blow this up; and make sure the caller has already done convertRightBuildersToIndex.